lib/spack/spack/build_environment.py

```diff
@@ -1437,3 +1424,51 @@ def write_log_summary(out, log_type, log, last=None):
         # If no errors are found but warnings are, display warnings
         out.write("\n%s found in %s log:\n" % (plural(nwar, "warning"), log_type))
         out.write(make_log_context(warnings))
+
+
+class ModuleMonkeyPatcher:
+    """Wrapper class to accept changes to a package.py Python module, and propagate them in the
+    MRO of the package.
+
+    It is mainly used as a substitute of the ``package.py`` module, when calling the
+    "setup_dependent_package" function during build environment setup.
+    """
+
+    _PROTECTED_NAMES = ("package", "current_module", "modules_in_mro", "_set_attributes")
+
+    def __init__(self, package):
+        self._set_self_attributes("package", package)
+        self._set_self_attributes("current_module", package.module)
+
+        #: Modules for the classes in the MRO up to PackageBase
+        modules_in_mro = []
+        for cls in inspect.getmro(type(package)):
+            module = cls.module
+
+            if module == self.current_module:
+                continue
+
+            if module == spack.package_base:
+                break
+
+            modules_in_mro.append(module)
+        self._set_self_attributes("modules_in_mro", modules_in_mro)
+        self._set_self_attributes("_set_attributes", {})
+
+    def _set_self_attributes(self, key, value):
+        super().__setattr__(key, value)
+
+    def __getattr__(self, item):
+        return getattr(self.current_module, item)
```

Comment:
I think this has a version of the same problem that the original PR was trying to fix -- we need to be able to access values from any of the higher modules in the class's MRO.

Could we do something like

```suggestion
    module_undefined = object()

    def __getattr__(self, item):
        for mod in [self.current_module] + self.modules_in_mro:
            value = getattr(mod, item, module_undefined)
            if value != self.module_undefined:
                return value

        getattr(self.current_module, item)  # raises attribute error with appropriate formatting
```

Reply:
I'm kind of opposed to this monkey patching also when getting objects. The receiver can't expect any order of execution for the `setup_dependent_package` functions, and the objects that might be queried are coming from there, right?

Reply:
To give an example, consider one file:
```python
SOME_GLOBAL = 1

class BaseFoo(MakefilePackage):
    [ ... ]
```
another file:
```python
class DerivedFoo(BaseFoo):
    [ ... ]
```
I don't want to allow retrieving:
```
derived_module.SOME_GLOBAL
```
since that is explicitly defined elsewhere. I think it's different for global variables that we inject in a controlled manner, like the ones in `setup_dependent_package`. In that case we want them to be propagated because a recipe might call `super()` and expect some function etc. to be there.

Reply:
By now I'm thinking it would be nicer to just pass a separate object of accumulated keys + values to the build environment, like we already do for environment variables. All this mutable state and global variables is just so ugly, hard to debug, hard to test, and nobody knows the order of evaluation.

Reply:
Here's the example I was thinking of:

```
# builtin repo
SOME_GLOBAL = 1
class Foo(Package):
    depends_on("bar")
    ...
    
# builtin repo
class Bar(Package):
    def setup_dependent_package(self, module, dependent_spec):
        if hasattr(module, 'SOME_GLOBAL'):
            ...

# local repo
class Foo(builtin.Foo):
    ...
```

In that case the module containing the class `Foo` has been effectively turned into part of the interface for `Foo`, but the inheriting package would have to reproduce it to be able to interoperate with the dependencies.

Reply:
@becker33 This is on the packager to fix. They can do:
```
# builtin repo
class Foo(Package):
    SOME_GLOBAL = 1

    depends_on("bar")
    ...
    
# builtin repo
class Bar(Package):
    def setup_dependent_package(self, module, dependent_spec):
        if hasattr(dependent_spec.package, 'SOME_GLOBAL'):
            ...

# local repo
class Foo(builtin.Foo):
    ...
```
which holds also for derived classes.

Reply:
I mean, we should probably be requiring that for everything and getting the module scopes out of the interface. But as long as we allow the builtin packages to have that relationship, we have to support this in the inherited package. The whole point of the inherited package is that you don't have to modify the builtin repo to be able to inherit from it.

Reply:
> By now I'm thinking it would be nicer to just pass a separate object of accumulated keys + values to the build environment, like we already do for environment variables. All this mutable state and global variables is just so ugly, hard to debug, hard to test, and nobody knows the order of evaluation.

I'd agree, but since this is part of the package interface we have the drawback of asking users to be more verbose. Instead of:
```python
cmake(...)
```
they have to do something similar to:
```python
commands["cmake"](...)
```
I am not sure if that might cause some push back. Also we need to be able to pass this context around to base classes etc.

Reply:
> But as long as we allow the builtin packages to have that relationship, we have to support this in the inherited package.

Here it's not the inherited package, but the inherited module. We shouldn't be looking into other modules and pretend the variables are defined in this module. We could find all sort of stuff including Builders etc. Also, I didn't try it, but is the current Spack behaving this way?

Reply:
We should document that any module-scope variables that are part of the package interface (used by a dependency/dependent) should be imported into the inheriting package, but otherwise I'm ok with that as a solution. I think we can expect that level of sophistication from users who are inheriting packages.